File: backend-data/backend/app/models/observability.py
"""统一链路日志 ORM，仅允许 backend-data 使用。"""

from datetime import datetime
from typing import Any
from uuid import UUID, uuid4

from sqlalchemy import (
    BigInteger,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class TraceRecordModel(Base):
    """链路摘要。"""

    __tablename__ = "trace_records"

    trace_id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), primary_key=True)
    trigger: Mapped[str] = mapped_column(String(40), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(300), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, index=True)
    started_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        index=True,
    )
    ended_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    duration_ms: Mapped[int] = mapped_column(BigInteger, nullable=False)
    root_service: Mapped[str] = mapped_column(String(40), nullable=False, index=True)
    http_method: Mapped[str | None] = mapped_column(String(16))
    http_path: Mapped[str | None] = mapped_column(String(500), index=True)
    http_status: Mapped[int | None] = mapped_column(Integer)
    error_message: Mapped[str | None] = mapped_column(Text)


class TraceSpanModel(Base):
    """链路树节点。"""

    __tablename__ = "trace_spans"
    __table_args__ = (
        Index("ix_trace_spans_trace_started", "trace_id", "started_at"),
    )

    span_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        default=uuid4,
    )
    trace_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("trace_records.trace_id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    parent_span_id: Mapped[UUID | None] = mapped_column(
        PG_UUID(as_uuid=True),
        nullable=True,
        index=True,
    )
    service: Mapped[str] = mapped_column(String(40), nullable=False, index=True)
    kind: Mapped[str] = mapped_column(String(20), nullable=False)
    operation: Mapped[str] = mapped_column(String(500), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, index=True)
    started_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    ended_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    duration_ms: Mapped[int] = mapped_column(BigInteger, nullable=False)
    attributes: Mapped[dict[str, Any]] = mapped_column(
        JSONB,
        nullable=False,
        default=dict,
    )
    error_message: Mapped[str | None] = mapped_column(Text)


class TraceEventModel(Base):
    """Span 内事件。"""

    __tablename__ = "trace_events"
    __table_args__ = (
        Index("ix_trace_events_trace_occurred", "trace_id", "occurred_at"),
    )

    event_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        default=uuid4,
    )
    trace_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("trace_records.trace_id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    span_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("trace_spans.span_id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    service: Mapped[str] = mapped_column(String(40), nullable=False, index=True)
    event_type: Mapped[str] = mapped_column(String(40), nullable=False, index=True)
    level: Mapped[str] = mapped_column(String(20), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(300), nullable=False)
    occurred_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        index=True,
    )
    attributes: Mapped[dict[str, Any]] = mapped_column(
        JSONB,
        nullable=False,
        default=dict,
    )


class TracePayloadModel(Base):
    """完整业务载荷元数据。"""

    __tablename__ = "trace_payloads"
    __table_args__ = (
        Index("ix_trace_payloads_trace_created", "trace_id", "created_at"),
    )

    payload_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        default=uuid4,
    )
    trace_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("trace_records.trace_id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    span_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("trace_spans.span_id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    service: Mapped[str] = mapped_column(String(40), nullable=False, index=True)
    payload_type: Mapped[str] = mapped_column(String(40), nullable=False, index=True)
    content_type: Mapped[str] = mapped_column(String(200), nullable=False)
    content_preview: Mapped[str] = mapped_column(Text, nullable=False, default="")
    content_sha256: Mapped[str] = mapped_column(String(64), nullable=False)
    chunk_count: Mapped[int] = mapped_column(Integer, nullable=False)
    size_bytes: Mapped[int] = mapped_column(BigInteger, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        index=True,
    )


class TracePayloadChunkModel(Base):
    """按 UTF-8 边界拆分的载荷正文。"""

    __tablename__ = "trace_payload_chunks"

    payload_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("trace_payloads.payload_id", ondelete="CASCADE"),
        primary_key=True,
    )
    chunk_index: Mapped[int] = mapped_column(Integer, primary_key=True)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    size_bytes: Mapped[int] = mapped_column(Integer, nullable=False)

File: frontend/src/pages/system/log-query/types/observability.ts
export type CallStatus = 'success' | 'warning' | 'failure';

export interface TraceRecord {
  trace_id: string;
  trigger: string;
  name: string;
  status: string;
  call_status: CallStatus;
  started_at: string;
  ended_at: string;
  duration_ms: number;
  root_service: string;
  http_method?: string | null;
  http_path?: string | null;
  http_status?: number | null;
  error_message?: string | null;
}

export interface TraceSpan {
  span_id: string;
  trace_id: string;
  parent_span_id?: string | null;
  service: string;
  kind: string;
  operation: string;
  status: string;
  started_at: string;
  ended_at: string;
  duration_ms: number;
  attributes: Record<string, unknown>;
  error_message?: string | null;
}

export interface TraceEvent {
  event_id: string;
  trace_id: string;
  span_id: string;
  service: string;
  event_type: string;
  level: string;
  name: string;
  occurred_at: string;
  attributes: Record<string, unknown>;
}

export interface TracePayload {
  payload_id: string;
  trace_id: string;
  span_id: string;
  service: string;
  payload_type: string;
  content_type: string;
  content_preview: string;
  content_sha256: string;
  chunk_count: number;
  size_bytes: number;
  created_at: string;
}

export interface TracePayloadChunk {
  payload_id: string;
  chunk_index: number;
  content: string;
  size_bytes: number;
}

export interface PaginatedPayloadChunks {
  items: TracePayloadChunk[];
  total: number;
  chunk_from: number;
  chunk_limit: number;
}

export interface PaginatedTraces {
  items: TraceRecord[];
  total: number;
  page: number;
  page_size: number;
}

export interface TraceDetail {
  trace: TraceRecord;
  spans: TraceSpan[];
  events: TraceEvent[];
}

export interface ObservabilityMetadata {
  triggers: string[];
  services: string[];
  span_kinds: string[];
  statuses: string[];
  levels: string[];
  event_types: string[];
  payload_types: string[];
  call_statuses: CallStatus[];
}

export interface TraceQuery {
  traceId?: string;
  startedFrom?: string;
  startedTo?: string;
  trigger?: string;
  service?: string;
  callStatus?: CallStatus;
  keyword?: string;
  page: number;
  pageSize: number;
}
